Add support for user-mode networking on Hyperv using cloud-init

lstocchi opened the pull request:
if usermode networking is enabled on hyperv we need to inject
gvforwarder and set up the vsock connection and a systemd service to
handle it.

To do this we create a custom user-data having all the data to inject in
the guest.

The resulting user-data is similar to https://github.com/crc-org/macadam/issues/228#issuecomment-3210712139

lstocchi committed:
move GenerateUserData to os specific impl

to support user-mode networking in hyperv we need to customize the
cloud-init config. This patch just move GenerateUserData outside of the
default cloudinit impl so it can be customized for hyperv in following
commit.

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
extend user-data if usermode networking is enabled in hyperv

if usermode networking is enabled on hyperv we need to inject
gvforwarder and set up the vsock connection and a systemd service to
handle it.

To do this we add gvforwarder binary within the ISO and use a custom user-data to copy
it in the right place in the guest.

The resulting user-data is similar to https://github.com/crc-org/macadam/issues/228#issuecomment-3334439056

Signed-off-by: lstocchi <lstocchi@redhat.com>

cfergeau commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -27,6 +33,61 @@ func GenerateUserData(mc *vmconfigs.MachineConfig) ([]byte, error) {
+		userData.WriteFiles = []WriteFile{
+			{
+				Path:        "/usr/local/bin/gvforwarder",
+				Content:     gvforwarderBase64,
```
Gerard expressed concerns about this approach, as this makes the cloud-init file huge. This might be ok with the way we pass the cloud-init data to the VM, but this can be problematic on other platforms, see https://github.com/vmware-tanzu/vm-operator/issues/138 for example.

lstocchi replied:
The alternative could be to add gvforwarder in the ISO we create having user-data so that cloud-init can move and use it. Or did you think at something else?

lstocchi commented:
@cfergeau code has been updated. Now it adds gvforwarder inside the ISO and then use the user-data to mount/copy the binary in the right location.

anjannath commented on pkg/machine/cloudinit/cloudinit.go:
```diff
@@ -138,6 +122,15 @@ func GenerateISO(mc *vmconfigs.MachineConfig) (*define.VMFile, error) {
 		return nil, err
 	}
 
+	resources := GetEmbeddedResources()
```
maybe we can add a check to see if a user provided cloud-init `user-data` is used, in that case we can skip putting the `gvproxy` in the ISO, wdyt?

lstocchi replied:
Yes, we can definitely add a check. Maybe it would be better to check if usermodeNetworking is enabled? If not, we skip it. 

I think we could try to add our own cloud-init config file to enable user-mode networking even if the user send his own user-data file.
E.g. if the user executes `macadam .... --user-mode-networking --cloud-init=my-user-data` we take his user-data and merge it with our user-data file to set up gvforwarder.

lstocchi replied:
@anjannath updated to support the idea exposed above. 
Let me know what you think. The `GetEmbeddedResources` contains the logic to skip it if there is nothing to be added.
It should work this way

1. **custom user-data + system networking** = we use user's user-data without any change and do not add any extra file to the iso

2. **NO custom user-data + system networking** = we generate a default user-data having the core user/ssh set (no extra file is added to the iso)

3. **custom user-data + user-mode networking** = we merge user's user-data with one created by us that contains what's needed to set/run gvforwarder + add gvforwarder binary in iso

4. **NO custom user-data + user-mode networking** = we generate a default user-data having the core user/ssh set and gvforwarder stuff + gvforwarder binary in iso

anjannath replied:
> custom user-data + user-mode networking = we merge user's user-data with one created by us that contains what's needed to set/run gvforwarder + add gvforwarder binary in iso

so the above scenario is what we'll have when using the self sufficient bundle with `macadam`, we'd provide a custom `user-data` and use the `user-mode` networking, but then the self sufficient bundle already has the `gvproxy` binary and the systemd service needed for user-mode, so if it is added again maybe things will not work as expected (need to test, i haven't tried running it with this PR yet)

the way this is added now is more suited to the use case of macadam as a generic tool to start VMs, but with the SNC self sufficient bundle we have a special case, so we have to find a way to handle it

anjannath replied:
maybe the `user-data` part that is added for hyperv + custom user-data + user-mode networking, the `run_cmd` part can check if the `gvisor-tap-vsock@tap0` service already exists its noop, otherwise it installs the binary and activate the service

something like:
```
if [ ! -f /etc/systemd/system/gvisort-tap-vsock@.service ]; then
  "install -o root -g root -m 0755 /mnt/gvforwarder /usr/local/bin/gvforwarder",
  "nmcli connection reload",
  "systemctl daemon-reload",
  "systemctl enable --now vsock-network.service",
fi
```

lstocchi replied:
this is something i did not try. I should verify it but on self sufficient bundle if we use the same systemd service/location for gvforwarder we should be fine. In the multipart cloud-init config it first adds our user-data, then the user's one. So cloud-init will execute our user-data, then yours. What i expect is it to try overwriting stuff but nothing should change, no?

lstocchi replied:
just read your new comment ... ah the systemd service is inside the self-sufficient bundle. You don't inject it

anjannath replied:
yeah this is already part of the bundle, is added during bundle generation here: https://github.com/crc-org/snc/blob/216b8312d00ec65ce3b848fb8c23612887eb9c77/createdisk.sh#L107

anjannath replied:
> this is something i did not try. I should verify it but on self sufficient bundle if we use the same systemd service/location for gvforwarder we should be fine.

i think this will also work, it'll need change on the SNC repo, but both creating in the same location with same content, so behavior should be same

lstocchi replied:
the other idea is if someone provides a network-config cloud-init file we skip this part as we assume they are setting networking themselves. And for the self sufficient bundle you just provide an empty network-config so cloud-init does not do anything but we skip to set up the systemd service/gvforwarder with macadam

lstocchi replied:
even though i'm not sure the network-config is really related to user-mode networking, so not sure it's conceptually correct. Need to check what users can do with that file

lstocchi replied:
@cfergeau any suggestion? ^^

cfergeau replied:
Imo trying to merge stuff we need with user-provided data will most likely have several unforeseen corner cases which might be hard to support, so I’d do that as little as possible.
However, another idea opposite to that which just came to mind, what if we removed `gvforwarder` and the related files from the self sufficient bundle, and relied on macadam to inject them when needed?
This would help with some of the issues @kpouget has been seeing.

lstocchi replied:
@anjannath for easier review i removed the multi-part user-data commit and open a new PR with it. If you want to test it you can use https://github.com/cfergeau/podman/pull/36 :)

cfergeau commented on pkg/machine/cloudinit/cloudinit_unix.go:
```diff
@@ -0,0 +1,40 @@
+//go:build !windows
```
The file could be named `cloudinit_unix.go`, which is an existing build constraint https://pkg.go.dev/cmd/go#hdr-Build_constraints
I’ve also seen `cloudinit_other.go` as far as naming go.
podman uses both.
`_rest.go` is confusing, I was wondering if this was about a REST HTTP API for cloud-init.

cfergeau commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -27,6 +32,55 @@ func GenerateUserData(mc *vmconfigs.MachineConfig) ([]byte, error) {
+			},
+			{
+				Path: "/etc/systemd/system/vsock-network.service",
+				Content: fmt.Sprintf(`[Unit]
```
Could this reuse `createNetworkUnit` from `pkg/machine/hyperv/stubber.go`?
I’m worried these 2 versions of the same file could go out of sync.

cfergeau commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -27,6 +32,55 @@ func GenerateUserData(mc *vmconfigs.MachineConfig) ([]byte, error) {
+		userData.WriteFiles = []WriteFile{
+			{
+				Path: "/etc/NetworkManager/system-connections/vsock0.nmconnection",
+				Content: `[connection]
```
Can this reuse `hyperVVsockNMConnection` from `pkg/machine/hyperv/stubber.go`?

cfergeau commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -38,3 +92,81 @@ func GenerateUserData(mc *vmconfigs.MachineConfig) ([]byte, error) {
+	}
+
+	for _, asset := range releaseInfo.Assets {
+		if asset.Name == "gvforwarder" {
```
NB: this wil need to be adjusted if one day we add support windows for arm

cfergeau commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -38,3 +92,81 @@ func GenerateUserData(mc *vmconfigs.MachineConfig) ([]byte, error) {
+		return nil, fmt.Errorf("failed to download gvforwarder: %s", resp.Status)
+	}
+
+	bytes, err := io.ReadAll(resp.Body)
```
This will allocate 5MB of RAM to store gvforwarder content, and will be turned back into an io.Reader to write it to the ISO, so it’s a bit wasteful.
We can improve this later though, and keep this code for the initial iteration.

cfergeau commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -27,6 +32,55 @@ func GenerateUserData(mc *vmconfigs.MachineConfig) ([]byte, error) {
 		},
 	}
 
+	if mc.HyperVHypervisor != nil && mc.HyperVHypervisor.UserModeNetworking {
```
Instead of duplicating `GenerateUserData`, maybe we could keep it in `cloudinit.go`, but add an OS specific `addProviderSpecificCloudInitData(…)` helper which `GenerateUserData` would call? (helper name can be improved, I used a long name for clarity in this comment)
This is similar to what you did in `GenerateUserDataFile`/`GetEmbeddedResources`

lstocchi replied:
I can think about doing something like

```
func generateUserData(mc *vmconfigs.MachineConfig) 
	userDataBytes := mc.CloudInitConfig.UserData
	if userDataBytes == nil {
		internalUserData, err = getDefaultUserData(mc)
		if err != nil {
			return nil, err
		}
		
		userDataBytes, err := yaml.Marshal(internalUserData)
		if err != nil {
			return nil, err
		}

		headerLine := "#cloud-config\n"
		userDataBytes= append([]byte(headerLine), userDataBytes...)
	}
	
	addProviderSpecificCloudInitData(userDataBytes, mc)	
```

but this way we always need to use MIME multi-part archive in `addProviderSpecificCloudInitData` even if we use our default user-data. Or were you thinking some other thing?

P.S: renamed `GenerateUserData` to `generateUserData`, as there was no reason to keep the capital

cfergeau reviewed (commented):
Overall looks good, however, we’ll also need an installer shipping all the binaries we require as we cannot rely on being able to download a file from the net, this could be blocked in corporate environments. This is also worrying from a security perspective as the binary has no signature/checksum/… check to try to ensure it can be trusted.

We can download it for now, and rework it later.

lstocchi committed:
update visibility createNetworkUnit and hyperVVsockNMConnection in hyperv stubber

this commit updates the visibility of createNetworkUnit and
hyperVVsockNMConnection and move it to another package so they can be reused
by the cloudinit package in further commit.

Signed-off-by: lstocchi <lstocchi@redhat.com>

cfergeau commented on pkg/machine/hyperv/hutil/hutil.go:
```diff
@@ -0,0 +1,35 @@
+package hutil
```
Is the new package required for circular deps? or you preferred to do it this way?

lstocchi replied:
yes, circular deps. So i copied what they did on wsl

cfergeau commented:
Maybe we could split the "multipart" commit to its own PR so that we can move forward with the others? I’ll need a bit longer to review it.

lstocchi commented:
> Maybe we could split the "multipart" commit to its own PR so that we can move forward with the others? I’ll need a bit longer to review it.

Removed it and open https://github.com/cfergeau/podman/pull/36
